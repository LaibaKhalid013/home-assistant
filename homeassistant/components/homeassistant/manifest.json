{
  "domain": "homeassistant",
  "name": "Home Assistant Core Integration",
  "after_dependencies": ["repairs"],
  "codeowners": ["@home-assistant/core"],
  "documentation": "https://www.home-assistant.io/integrations/homeassistant",
  "integration_type": "system",
  "quality_scale": "internal"
}
