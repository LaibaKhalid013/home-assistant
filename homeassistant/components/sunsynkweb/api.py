"""stub for api access functions not attached to objects."""

import logging
import pprint

from .const import BASE_HEADERS, BASE_URL

_LOGGER = logging.getLogger(__name__)


async def get_bearer_token(session, username, password):
    """Get the bearer token for the sunsynk api."""
    params = {
        "username": username,
        "password": password,
        "grant_type": "password",
        "client_id": "csp-web",
        "source": "sunsynk",
        "areaCode": "sunsynk",
    }
    returned = await session.post(
        BASE_URL + "/oauth/token", json=params, headers=BASE_HEADERS
    )

    returned = await returned.json()
    _LOGGER.debug("authentication attempt returned %s", pprint.pformat(returned))
    # returned data looks like the below
    # {
    #     "code": 0,
    #     "msg": "Success",
    #     "data": {
    #         "access_token": "VALUE",
    #         "token_type": "bearer",
    #         "refresh_token": "VALUE",
    #         "expires_in": 604799,
    #         "scope": "all",
    #     },
    #     "success": True,
    # }
    return returned["data"]["access_token"]
