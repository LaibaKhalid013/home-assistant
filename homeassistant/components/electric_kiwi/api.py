"""API for Electric Kiwi bound to Home Assistant OAuth."""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

from electrickiwi_api import AbstractAuth

from .const import API_BASE_URL

if TYPE_CHECKING:
    from aiohttp import ClientSession

    from homeassistant.helpers import config_entry_oauth2_flow


class AsyncConfigEntryAuth(AbstractAuth):
    """Provide Electric Kiwi authentication tied to an OAuth2 based config entry."""

    def __init__(
        self,
        websession: ClientSession,
        oauth_session: config_entry_oauth2_flow.OAuth2Session,
    ) -> None:
        """Initialize Electric Kiwi auth."""
        # add host when ready for production "https://api.electrickiwi.co.nz" defaults to dev
        super().__init__(websession, API_BASE_URL)
        self._oauth_session = oauth_session

    async def async_get_access_token(self) -> str:
        """Return a valid access token."""
        if not self._oauth_session.valid_token:
            await self._oauth_session.async_ensure_token_valid()

        return cast(str, self._oauth_session.token["access_token"])
