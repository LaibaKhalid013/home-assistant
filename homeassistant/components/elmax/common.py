"""Elmax integration common classes and utilities."""

from __future__ import annotations

import ssl

from elmax_api.model.endpoint import DeviceEndpoint
from elmax_api.model.panel import PanelEntry
from packaging import version

from homeassistant.core import callback
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.dispatcher import async_dispatcher_connect
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import (
    DOMAIN,
    ELMAX_LOCAL_API_PATH,
    MIN_APIV2_SUPPORTED_VERSION,
    SIGNAL_PANEL_UPDATE,
)
from .coordinator import ElmaxCoordinator


def get_direct_api_url(host: str, port: int, use_ssl: bool) -> str:
    """Return the direct API url given the base URI."""
    schema = "https" if use_ssl else "http"
    return f"{schema}://{host}:{port}/{ELMAX_LOCAL_API_PATH}"


def build_direct_ssl_context(cadata: str) -> ssl.SSLContext:
    """Create a custom SSL context for direct-api verification."""
    context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_CLIENT)
    context.check_hostname = False
    context.verify_mode = ssl.CERT_REQUIRED
    context.load_verify_locations(cadata=cadata)
    return context


def check_local_version_supported(api_version: str | None) -> bool:
    """Check whether the given API version is supported."""
    if api_version is None:
        return False
    return version.parse(api_version) >= version.parse(MIN_APIV2_SUPPORTED_VERSION)


class DirectPanel(PanelEntry):
    """Helper class for wrapping a directly accessed Elmax Panel."""

    def __init__(self, panel_uri):
        """Construct the object."""
        super().__init__(panel_uri, True, {})

    def get_name_by_user(self, username: str) -> str:
        """Return the panel name."""
        return f"Direct Panel {self.hash}"


class ElmaxEntity(CoordinatorEntity[ElmaxCoordinator]):
    """Wrapper for Elmax entities."""

    _last_state: DeviceEndpoint = None

    def __init__(
        self,
        elmax_device: DeviceEndpoint,
        panel_version: str,
        coordinator: ElmaxCoordinator,
    ) -> None:
        """Construct the object."""
        super().__init__(coordinator=coordinator)
        self._device = elmax_device
        self._attr_unique_id = elmax_device.endpoint_id
        self._attr_name = elmax_device.name
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, coordinator.panel_entry.hash)},
            name=coordinator.panel_entry.get_name_by_user(
                coordinator.http_client.get_authenticated_username()
            ),
            manufacturer="Elmax",
            model=panel_version,
            sw_version=panel_version,
        )

    async def async_added_to_hass(self) -> None:
        """Register push notifications callbacks if available."""
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass,
                f"{SIGNAL_PANEL_UPDATE}-{self.coordinator.panel_entry.hash}-{self._device.endpoint_id}",
                self._handle_update,
            )
        )
        return await super().async_added_to_hass()

    @callback
    def _handle_update(self, endpoint_status: DeviceEndpoint) -> None:
        self._last_state = endpoint_status
        self.async_write_ha_state()

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return super().available and self.coordinator.panel_entry.online
