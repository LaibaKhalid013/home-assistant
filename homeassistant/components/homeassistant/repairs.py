"""Repairs for Home Assistant."""

from __future__ import annotations

from homeassistant import data_entry_flow
from homeassistant.components.repairs import ConfirmRepairFlow, RepairsFlow
from homeassistant.core import HomeAssistant


class IntegrationNotFoundFlow(RepairsFlow):
    """Handler for an issue fixing flow."""

    def __init__(self, data: dict[str, str]) -> None:
        """Initialize."""
        self.description_placeholders: dict[str, str] = {
            "device_name": data["device_name"]
        }

    async def async_step_init(
        self, user_input: dict[str, str] | None = None
    ) -> data_entry_flow.FlowResult:
        """Handle the first step of a fix flow."""
        return self.async_show_menu(
            menu_options=["confirm", "ignore"],
            description_placeholders=self.description_placeholders,
        )


async def async_create_fix_flow(
    hass: HomeAssistant, issue_id: str, data: dict[str, str] | None
) -> RepairsFlow:
    """Create flow."""

    if issue_id.split(".")[0] == "integration_not_found":
        assert data
        return IntegrationNotFoundFlow(data)
    return ConfirmRepairFlow()
