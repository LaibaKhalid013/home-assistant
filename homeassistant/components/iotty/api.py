"""API for iotty bound to Home Assistant OAuth."""

from __future__ import annotations

import logging
from typing import Any

from aiohttp import ClientSession
from iottycloud.cloudapi import CloudApi

from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_entry_oauth2_flow

from .const import IOTTYAPI_BASE, OAUTH2_CLIENT_ID

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [Platform.SWITCH]


class IottyProxy(CloudApi):
    """Provide iotty authentication tied to an OAuth2 based config entry."""

    def __init__(
        self,
        hass: HomeAssistant,
        websession: ClientSession,
        oauth_session: config_entry_oauth2_flow.OAuth2Session,
    ) -> None:
        """Initialize iotty auth."""

        super().__init__(websession, IOTTYAPI_BASE, OAUTH2_CLIENT_ID)
        if oauth_session is None:
            raise ValueError("oauth_session")
        self._oauth_session = oauth_session
        self._hass = hass

    async def async_get_access_token(self) -> Any:
        """Return a valid access token."""

        if not self._oauth_session.valid_token:
            await self._oauth_session.async_ensure_token_valid()

        return self._oauth_session.token["access_token"]
