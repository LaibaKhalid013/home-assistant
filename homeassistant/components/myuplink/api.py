"""API for myUplink bound to Home Assistant OAuth."""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

from myuplink.auth_abstract import AbstractAuth

from .const import API_ENDPOINT

if TYPE_CHECKING:
    from aiohttp import ClientSession

    from homeassistant.helpers import config_entry_oauth2_flow


class AsyncConfigEntryAuth(AbstractAuth):
    """Provide myUplink authentication tied to an OAuth2 based config entry."""

    def __init__(
        self,
        websession: ClientSession,
        oauth_session: config_entry_oauth2_flow.OAuth2Session,
    ) -> None:
        """Initialize myUplink auth."""
        super().__init__(websession, API_ENDPOINT)
        self._oauth_session = oauth_session

    async def async_get_access_token(self) -> str:
        """Return a valid access token."""
        if not self._oauth_session.valid_token:
            await self._oauth_session.async_ensure_token_valid()

        return cast(str, self._oauth_session.token["access_token"])
