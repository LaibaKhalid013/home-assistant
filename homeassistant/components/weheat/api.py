"""API for Weheat bound to Home Assistant OAuth."""

# import my_pypi_package
from abc import ABC, abstractmethod

from aiohttp import ClientResponse, ClientSession

from homeassistant.helpers import config_entry_oauth2_flow

from .const import API_URL


# this has to move to the backend client package
class AbstractAuth(ABC):
    """Abstract class to make authenticated requests."""

    def __init__(self, websession: ClientSession, host: str) -> None:
        """Initialize the auth."""
        self.websession = websession
        self.host = host

    @abstractmethod
    async def async_get_access_token(self) -> str:
        """Return a valid access token."""

    async def request(self, method, url, **kwargs) -> ClientResponse:
        """Make a request."""
        headers = kwargs.get("headers")

        if headers is None:
            headers = {}
        else:
            headers = dict(headers)

        access_token = await self.async_get_access_token()
        headers["authorization"] = f"Bearer {access_token}"

        return await self.websession.request(
            method,
            f"{self.host}/{url}",
            **kwargs,
            headers=headers,
        )


class AsyncConfigEntryAuth(AbstractAuth):
    """Provide Weheat authentication tied to an OAuth2 based config entry."""

    def __init__(
        self,
        websession: ClientSession,
        oauth_session: config_entry_oauth2_flow.OAuth2Session,
    ) -> None:
        """Initialize Weheat auth."""
        super().__init__(websession, host=API_URL)
        self._oauth_session = oauth_session

    async def async_get_access_token(self) -> str:
        """Return a valid access token."""
        if not self._oauth_session.valid_token:
            await self._oauth_session.async_ensure_token_valid()

        return self._oauth_session.token["access_token"]
