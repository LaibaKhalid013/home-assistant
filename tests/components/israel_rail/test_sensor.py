"""Tests for the israel_rail sensor."""

from __future__ import annotations

from unittest.mock import AsyncMock

from freezegun.api import FrozenDateTimeFactory

from homeassistant.core import HomeAssistant

from . import goto_future, init_integration
from .conftest import trains, trains_wrong_format

from tests.common import MockConfigEntry


async def test_valid_config(
    hass: HomeAssistant,
    mock_israelrail: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Ensure everything starts correctly."""
    await init_integration(hass, mock_config_entry)
    assert len(hass.states.async_entity_ids()) == 7
    assert hass.states.get("sensor.mock_title_departure")
    assert hass.states.get("sensor.mock_title_departure_1")
    assert hass.states.get("sensor.mock_title_departure_2")
    assert hass.states.get("sensor.mock_title_duration")
    assert hass.states.get("sensor.mock_title_transfers")
    assert hass.states.get("sensor.mock_title_platform")
    assert hass.states.get("sensor.mock_title_train_number")


async def test_invalid_config(
    hass: HomeAssistant, mock_config_entry: MockConfigEntry
) -> None:
    """Ensure nothing is created when config is wrong."""
    await init_integration(hass, mock_config_entry)
    assert not hass.states.async_entity_ids("sensor")


async def test_update_train(
    hass: HomeAssistant,
    freezer: FrozenDateTimeFactory,
    mock_israelrail: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Ensure the train data is updated."""
    await init_integration(hass, mock_config_entry)
    assert len(hass.states.async_entity_ids()) == 7
    departure_sensor = hass.states.get("sensor.mock_title_departure")
    assert departure_sensor.state == "2021-10-10T07:10:10+00:00"

    mock_israelrail.return_value.query.return_value = trains[1:]

    await goto_future(hass, freezer)

    assert len(hass.states.async_entity_ids()) == 7
    departure_sensor = hass.states.get("sensor.mock_title_departure")
    assert departure_sensor.state == "2021-10-10T07:20:10+00:00"


async def test_no_duration_wrong_date_format(
    hass: HomeAssistant,
    mock_israelrail: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Ensure the duration is not set when there is no departure time."""
    mock_israelrail.return_value.query.return_value = trains_wrong_format
    await init_integration(hass, mock_config_entry)
    assert len(hass.states.async_entity_ids()) == 7
    departure_sensor = hass.states.get("sensor.mock_title_train_number")
    assert departure_sensor.state == "1234"
    duration_sensor = hass.states.get("sensor.mock_title_duration")
    assert duration_sensor.state == "unknown"


async def test_fail_query(
    hass: HomeAssistant,
    freezer: FrozenDateTimeFactory,
    mock_israelrail: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Ensure the integration handles query failures."""
    await init_integration(hass, mock_config_entry)
    assert len(hass.states.async_entity_ids()) == 7
    mock_israelrail.return_value.query.side_effect = Exception("error")
    await goto_future(hass, freezer)
    assert len(hass.states.async_entity_ids()) == 7
    departure_sensor = hass.states.get("sensor.mock_title_departure")
    assert departure_sensor.state == "unavailable"
