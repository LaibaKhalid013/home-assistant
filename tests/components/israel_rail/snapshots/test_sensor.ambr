# serializer version: 1
# name: test_valid_config[sensor.mock_title_departure]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Israel rail.',
      'destination': 'Tel Aviv-Savidor Center',
      'device_class': 'timestamp',
      'friendly_name': 'Mock Title Departure',
      'start': 'Hertsliya',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_departure',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '2021-10-10T07:10:10+00:00',
  })
# ---
# name: test_valid_config[sensor.mock_title_departure_1]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Israel rail.',
      'destination': 'Tel Aviv-Savidor Center',
      'device_class': 'timestamp',
      'friendly_name': 'Mock Title Departure +1',
      'start': 'Hertsliya',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_departure_1',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '2021-10-10T07:20:10+00:00',
  })
# ---
# name: test_valid_config[sensor.mock_title_departure_2]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Israel rail.',
      'destination': 'Tel Aviv-Savidor Center',
      'device_class': 'timestamp',
      'friendly_name': 'Mock Title Departure +2',
      'start': 'Hertsliya',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_departure_2',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '2021-10-10T07:30:10+00:00',
  })
# ---
# name: test_valid_config[sensor.mock_title_duration]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Israel rail.',
      'device_class': 'duration',
      'friendly_name': 'Mock Title Duration',
      'unit_of_measurement': <UnitOfTime.SECONDS: 's'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_duration',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1200',
  })
# ---
# name: test_valid_config[sensor.mock_title_platform]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Israel rail.',
      'friendly_name': 'Mock Title Platform',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_platform',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1',
  })
# ---
# name: test_valid_config[sensor.mock_title_train_number]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Israel rail.',
      'friendly_name': 'Mock Title Train Number',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_train_number',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1234',
  })
# ---
# name: test_valid_config[sensor.mock_title_transfers]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Israel rail.',
      'friendly_name': 'Mock Title Transfers',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_transfers',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '0',
  })
# ---
